server/service/schedule/schedule.go

```diff
@@ -0,0 +1,294 @@
+// Package schedule allows periodic run of a list of jobs.
+//
+// Type Schedule allows grouping a set of Jobs to run at specific intervals.
+// Each Job is executed serially in the order they were added to the Schedule.
+package schedule
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/getsentry/sentry-go"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+// ReloadInterval reloads and returns a new interval.
+type ReloadInterval func(ctx context.Context) (time.Duration, error)
+
+// Schedule runs a list of jobs serially at a given schedule.
+//
+// Each job is executed one after the other in the order they were added.
+// If one of the job fails, an error is logged and the scheduler
+// continues with the next.
+type Schedule struct {
+	ctx        context.Context
+	name       string
+	instanceID string
+	logger     log.Logger
+
+	schedIntervalMu sync.Mutex // protects schedInterval.
+	schedInterval   time.Duration
+
+	done chan struct{}
+
+	configReloadInterval   time.Duration
+	configReloadIntervalFn ReloadInterval
+
+	locker Locker
+
+	altLockName string
+
+	jobs []Job
+}
+
+// JobFn is the signature of a Job.
+type JobFn func(context.Context) error
+
+// Job represents a job that can be added to Scheduler.
+type Job struct {
```

Comment:
Addressing @mna's [comment](https://github.com/fleetdm/fleet/pull/5727#discussion_r876209746) from the prior PR:

> It seems like this would be more flexible if it was an interface instead, with ID() string and something like RunScheduledJob(context.Context) error. The way it is, you have to use a closure to capture the necessary dependencies of the job (as is done in the cron.go file), but if it was an interface jobs could be defined a bit cleaner, with a struct and fields to set the dependencies.

I've taken an initial stab at this suggestion and I'd appreciate feedback on the interface design as I'm sure it can be improved :) 

Although we may just if we should just keep things simple for now and go with the original function closure approach and consider expanding the interface down the road.

Reply:
On second thought, I'm leaning strongly toward keeping the original function closure approach for now so I reverted back.